### PR TITLE
Move resolve_instance to rustc_ty.

### PR DESCRIPTION
**Pull request by cjgillot**
r? @Zoxc

### CODE REVIEW

**Comment by Zoxc** [comment]
These don't really belong in `rustc_resolve`. I'd put them in `rustc_ty` instead.

File: src/librustc/ty/instance.rs [hunk 1]
```diff
@@ -512,3 +391,29 @@ fn needs_fn_once_adapter_shim(
...
+pub static RESOLVE_INSTANCE: AtomicRef<
+    for<'tcx> fn(
+        TyCtxt<'tcx>,
+        ty::ParamEnv<'tcx>,
+        DefId,
+        SubstsRef<'tcx>,
+    ) -> Option<Instance<'tcx>>,
+> = AtomicRef::new(
+    &(resolve_instance_default
+        as for<'tcx> fn(
```
Comment (Zoxc) [inline src/librustc/ty/instance.rs]: Can you do a `as _` cast here?

**Comment by cjgillot** [comment]
What should be done with the perf result?

**Comment by petrochenkov** [comment]
@cjgillot 
Could you squash the commits before this PR is merged, so it doesn't pollute `rustc_resolve` history?

**Comment by cjgillot** [comment]
Squashed.

**Comment by Zoxc** [comment]
The performance overhead will be fixed by https://github.com/rust-lang/rust/pull/67797 so I'm just going to approve this.

@bors r+ p=1 rollup=never

**Pull request merged**